### ISSUE TITLE
Support extensions in Route matching

### ISSUE THREAD

**Issue by agrafix** [agrafix/Spock#4]
Support routes like `"foo" <//> var <.> "html"`

**Comment by simon-nicholls** [comment]
This would be good to see. The above linked pull request looked straightforward, relative to the previous routing, but I'm not ready to stick my fingers into the gears of the newer reroute lib.

I guess middleware URL preprocessing is the way to go for now.

**Comment by agrafix** [comment]
Yes, this would have to go into the reroute lib. I'm still thinking about how to implement this nicely, because it might be cool to track the extension at typelevel, too.

**Comment by agrafix** [comment]
This will probably be easy to do once #84 is merged.

**Comment by simon-nicholls** [comment]
Looking at this for the next release then? That milestone change looks very promising! Routing is one of those things that I find tricky when trying to support old routes from other app stacks.

**Comment by agrafix** [comment]
@simon-nicholls I am looking into this (again), but you post some example routes? I'm thinking that you probably could already implement it with existing components.

**Comment by simon-nicholls** [comment]
A couple of paths like this crop up for me, where content-length client-side is restricted to 2048 bytes a *page*:

    /values/<key>/pages/<page>.txt

I'm using `Network.Wai.Middleware.Rewrite` to strip off the extension at the moment, meaning I can leave a trailing var in the route:

    get (var <//> "pages" <//> var) $ \k p -> do

which works, though access to the extension type could be important at some point.

Having a `var` to scoop up the extension as well in a `Var Text` and then manually processing further would work as a last resort I guess. I did try:

    get (var <//> "pages" <//> var <.> var) $ \k p e -> do

but it was wishful thinking. I've been noodling more seriously with Spock the last couple of days, and it's been great fun!

**Comment by moll** [comment]
For the record, I'm using an `IdWithExtension` `newtype` to achieve this. Might even be more extensible than adding extension support to Reroute. After all, there are a myriad of ways to encode things — multiple extensions, alternative syntax etc — that can't all go to Reroute.

```haskell
data IdWithExtension id = IdWithExtension id Text

instance (HttpApiData.FromHttpApiData id) => HttpApiData.FromHttpApiData (IdWithExtension id) where
  parseUrlPiece path = case Text.break (== '.') path of
      (_id, "") -> Left "No extension"
      (id, ext) -> IdWithExtension <$> HttpApiData.parseUrlPiece id <*> pure (Text.drop 1 ext)
```
